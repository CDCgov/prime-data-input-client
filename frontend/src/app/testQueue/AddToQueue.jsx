import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import { v4 as uuidv4 } from "uuid";

import Button from "../commonComponents/Button";
import SearchBar from "../commonComponents/SearchBar";
import { searchPatients } from "../query/patients";
import { displayFullName } from "../utils";

const AddToQueue = () => {
  let history = useHistory();

  const [patientSearchResults, updatePatientSearchResults] = useState([]);

  const onSearchClick = (event, searchQuery) => {
    event.preventDefault();
    let searchResults = searchPatients(searchQuery);
    updatePatientSearchResults(searchResults);
  };

  const renderSearchResults = () => {
    if (patientSearchResults.length === 0) {
      return null;
    }
    let rows = patientSearchResults.map((patient) => (
      <tr key={`patient-${uuidv4()}`}>
        <th scope="row">
          {displayFullName(
            patient.firstName,
            patient.middleName,
            patient.lastName
          )}
        </th>
        <td>{patient.patientId}</td>
        <td>{patient.birthDate}</td>
        <td>Add to Queue</td>
      </tr>
    ));
    return (
      <React.Fragment>
        <h1> Results: </h1>
        <table className="usa-table usa-table--borderless">
          <thead>
            <tr>
              <th scope="col">Full Name</th>
              <th scope="col">Unique ID</th>
              <th scope="col">Date of Birth</th>
              <th scope="col">Actions</th>
            </tr>
          </thead>
          <tbody>{rows}</tbody>
        </table>
      </React.Fragment>
    );
  };

  return (
    <main className="prime-home">
      <div className="grid-container">
        <div className="grid-row">
          <div className="grid-col">
            <Button
              icon="arrow-left"
              type="button"
              onClick={() => history.goBack()}
              label="  Go Back"
            />
            <div className="prime-container">
              <SearchBar onSearchClick={onSearchClick} />
              {renderSearchResults()}
            </div>
          </div>
        </div>
      </div>
    </main>
  );
};

export default AddToQueue;
