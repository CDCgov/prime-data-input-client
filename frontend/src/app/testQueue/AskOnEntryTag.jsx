import React from "react";

export const areAnswersComplete = (answerDict) => {
  if (!answerDict.noSymptoms) {
    let symptomFound = false;

    try {
      const symptoms = JSON.parse(answerDict.symptoms);
      Object.values(symptoms).forEach((val) => {
        if (val) {
          symptomFound = true;
        }
      });
    } catch (e) {
      console.error("expected json response. found:", e);
    }

    if (!symptomFound) {
      return false;
    }
    if (answerDict.symptomOnset) {
      const onsetDate = answerDict.symptomOnset;
      if (!onsetDate) {
        return false;
      }
    }
  }
  if (!answerDict.firstTest) {
    if (!answerDict.priorTestDate) {
      return false;
    }
    if (!answerDict.priorTestType || !answerDict.priorTestResult) {
      return false;
    }
  }
  if (!answerDict.pregnancy) {
    return false;
  }
  return true;
};

const AskOnEntryTag = ({ aoeAnswers }) => {
  if (areAnswersComplete(aoeAnswers)) {
    return <span className="usa-tag bg-green">COMPLETED</span>;
  } else {
    return <span className="usa-tag">PENDING</span>;
  }
};

export default AskOnEntryTag;
