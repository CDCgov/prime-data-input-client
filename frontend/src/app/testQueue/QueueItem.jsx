import React from "react";
import { useDispatch, useSelector } from "react-redux";

import { patientPropType } from "../propTypes";
import LabeledText from "../commonComponents//LabeledText";
import TestResultInputForm from "../testResults/TestResultInputForm";
import Dropdown from "../commonComponents//Dropdown";
import { getTestResultById } from "../testResults/testResultsSelector";
import Button from "../commonComponents/Button";
import { removePatientFromQueue } from "./state/testQueueActions";

const QueueItem = ({ patient }) => {
  const onSubmit = (e) => {
    e.preventDefault();
  };

  const onDropdownChange = (e) => {
    console.log(e.target.value);
  };

  const dispatch = useDispatch();

  const removeFromQueue = (patientId) => {
    dispatch(removePatientFromQueue(patientId));
  };

  // useEffect(() => {
  //   dispatch(loadPatients(organizationId));
  // }, [organizationId, dispatch]);

  const testResult = useSelector(getTestResultById(patient.testResultId));

  return (
    <React.Fragment>
      <div className="grid-container prime-container prime-queue-item">
        <Button
          icon="times-circle"
          onClick={() => removeFromQueue(patient.patientId)}
        />
        <div className="grid-row">
          <div className="tablet:grid-col-9">
            <div className="grid-row prime-test-name">
              <h1>
                {patient.firstName} {patient.lastName}
              </h1>
            </div>
            <div className="grid-row">
              <ul className="prime-ul">
                <li className="prime-li">
                  <LabeledText text={patient.patientId} label="Unique ID" />
                </li>
                <li className="prime-li">
                  <LabeledText text={patient.phone} label="Phone Number" />
                </li>
                <li className="prime-li">
                  <LabeledText text={patient.birthDate} label="Date of Birth" />
                </li>
              </ul>
            </div>
            <div className="grid-row">
              <Dropdown
                options={[
                  { text: "Abbott ID Now", value: "abbottIdNow" },
                  { text: "Some other device", value: "someOtherDeviceValue" },
                ]}
                label="Device"
                name="testDevice"
                onChange={onDropdownChange}
              />
            </div>
          </div>
          <div className="tablet:grid-col-3 prime-test-result prime-container-padding">
            <TestResultInputForm testResult={testResult} onSubmit={onSubmit} />
          </div>
        </div>
      </div>
    </React.Fragment>
  );
};

QueueItem.propTypes = {
  patient: patientPropType,
};
export default QueueItem;
