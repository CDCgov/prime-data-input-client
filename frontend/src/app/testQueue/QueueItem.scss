#patient-name-header {
  padding: 1em;
  padding-left: 2em;
}

.testdate-checkbox {
  top: 1.5rem;
  position: relative;
} 

.test-information-ul .usa-legend, .test-information-ul  .usa-label {
  display:inline;
  line-height: 1.6;
}

.prime-ul, .test-information-ul {
  margin:0rem; 
}