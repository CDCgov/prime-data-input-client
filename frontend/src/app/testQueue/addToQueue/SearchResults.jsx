import React from "react";
import { v4 as uuidv4 } from "uuid";
import { useDispatch } from "react-redux";
import PropTypes from "prop-types";

import Anchor from "../../commonComponents/Anchor";
import { addPatientToQueue } from "../state/testQueueActions";

const SearchResults = ({ suggestions, shouldDisplay, onAddToQueue }) => {
  const dispatch = useDispatch();

  if (!shouldDisplay) {
    return null;
  }

  const addToQueueButton = (patient) =>
    patient.isInQueue ? (
      "Already in queue"
    ) : (
      <Anchor
        onClick={() => {
          dispatch(addPatientToQueue(patient.patientId));
          onAddToQueue();
        }}
        text="Add to Queue"
      />
    );

  const renderResultsTable = () => {
    if (suggestions.length === 0) {
      return <h3> No results </h3>;
    }
    let suggestionRows = suggestions.map((suggestion) => (
      <tr key={uuidv4()}>
        <td>{suggestion.displayName}</td>
        <td>{suggestion.birthDate}</td>
        <td>{suggestion.patientId}</td>
        <td>{addToQueueButton(suggestion)}</td>
      </tr>
    ));
    return (
      <table className="usa-table usa-table--borderless">
        <thead>
          <tr>
            <th scope="col">Full name</th>
            <th scope="col">Date of Birth</th>
            <th scope="col">Unique ID</th>
            <th scope="col">Actions</th>
          </tr>
        </thead>
        <tbody>{suggestionRows}</tbody>
      </table>
    );
  };

  return (
    <div className="usa-card__container">
      <div className="usa-card__body">{renderResultsTable()}</div>
    </div>
  );
};

SearchResults.propTypes = {
  suggestions: PropTypes.arrayOf(PropTypes.any), // TODO: once this stabilizes define what `any` means
  shouldDisplay: PropTypes.bool,
  onAddToQueu: PropTypes.func,
};

export default SearchResults;
