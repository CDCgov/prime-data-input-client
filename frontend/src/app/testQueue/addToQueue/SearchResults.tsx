import React, { useState } from "react";
import Button from "../../commonComponents/Button";
import AoeModalForm from "../AoEForm/AoEModalForm";
import { displayFullName } from "../../utils";

interface SearchResultsProps {
  patients: any[]; //TODO TYPE: Patient
  onAddToQueue: (a: any, b: any, c: string) => string; //TODO TYPE: Patient, answers
  facilityId: string;
  patientsInQueue: string[];
  shouldShowSuggestions: boolean;
}

const SearchResults = ({
  patients,
  onAddToQueue,
  facilityId,
  patientsInQueue,
  shouldShowSuggestions,
}: SearchResultsProps) => {
  const [dialogPatient, setDialogPatient] = useState(null);
  const [canAddToQueue, setCanAddToQueue] = useState(false);

  const canAddToTestQueue = (patientId = "") => {
    return patientsInQueue.indexOf(patientId) === -1;
  };
  let results;
  if (patients.length === 0) {
    results = <h3> No results </h3>;
    if (canAddToQueue) setCanAddToQueue(false);
  } else {
    results = (
      <div className="usa-card__container shadow-3 results-dropdown">
        <div className="usa-card__body">
          <table className="usa-table usa-table--borderless">
            <thead>
              <tr>
                <th scope="col">Full name</th>
                <th scope="col">Date of Birth</th>
                <th scope="col">Unique ID</th>
                <th scope="col">Actions</th>
              </tr>
            </thead>
            <tbody>
              {patients.map((p) => (
                <tr key={p.internalId}>
                  <td>
                    {displayFullName(p.firstName, p.middleName, p.lastName)}
                  </td>
                  <td>{p.birthDate}</td>
                  <td>{p.lookupId}</td>
                  <td>
                    {canAddToTestQueue(p.internalId) ? (
                      <Button
                        variant="unstyled"
                        label="Begin Test"
                        onClick={() => {
                          setDialogPatient(p);
                          setCanAddToQueue(canAddToTestQueue(p.internalId));
                        }}
                      />
                    ) : (
                      "Test in progress"
                    )}
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>
    );
  }

  return (
    <>
      {dialogPatient !== null && (
        <AoeModalForm
          saveButtonText="Continue"
          patient={dialogPatient}
          onClose={() => {
            setDialogPatient(null);
          }}
          saveCallback={(a: any) =>
            onAddToQueue(dialogPatient, a, canAddToQueue ? "create" : "update")
          }
          facilityId={facilityId}
          canAddToTestQueue={canAddToQueue}
        />
      )}
      {shouldShowSuggestions && results}
    </>
  );
};

export default SearchResults;
