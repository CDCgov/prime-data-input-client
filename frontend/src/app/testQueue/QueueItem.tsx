import React, { useState } from "react";
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { toast } from "react-toastify";
import { gql, useMutation } from "@apollo/client";
import Modal from "react-modal";
import {
  useAppInsightsContext,
  useTrackEvent,
} from "@microsoft/applicationinsights-react-js";

import Alert from "../commonComponents/Alert";
import Anchor from "../commonComponents/Anchor";
import AoeModalForm from "./AoEForm/AoEModalForm";
import Dropdown from "../commonComponents/Dropdown";
import LabeledText from "../commonComponents/LabeledText";
import TestResultInputForm from "../testResults/TestResultInputForm";
import { ALERT_CONTENT } from "./constants";
import { displayFullName } from "../utils";
import { patientPropType, devicePropType } from "../propTypes";
import { QUEUE_NOTIFICATION_TYPES } from "./constants";
import { showNotification } from "../utils";
import AskOnEntryTag, { areAnswersComplete } from "./AskOnEntryTag";
import { removeTimer, TestTimerWidget } from "./TestTimer";
import moment from "moment";
import Button from "../commonComponents/Button";

export type TestResult = "POSITIVE" | "NEGATIVE" | "UNDETERMINED";

const REMOVE_PATIENT_FROM_QUEUE = gql`
  mutation RemovePatientFromQueue($patientId: String!) {
    removePatientFromQueue(patientId: $patientId)
  }
`;

const EDIT_QUEUE_ITEM = gql`
  mutation EditQueueItem($id: String!, $deviceId: String, $result: String) {
    editQueueItem(id: $id, deviceId: $deviceId, result: $result) {
      result
      deviceType {
        internalId
      }
    }
  }
`;

interface EditQueueItemParams {
  id: string;
  deviceId?: string;
  result?: TestResult;
}

interface EditQueueItemResponse {
  editQueueItem: {
    result: TestResult;
    deviceType: { internalId: string };
  };
}

const SUBMIT_TEST_RESULT = gql`
  mutation SubmitTestResult(
    $patientId: String!
    $deviceId: String!
    $result: String!
  ) {
    addTestResult(patientId: $patientId, deviceId: $deviceId, result: $result)
  }
`;

const UPDATE_AOE = gql`
  mutation UpdateAOE(
    $patientId: String!
    $symptoms: String
    $symptomOnset: String
    $pregnancy: String
    $firstTest: Boolean
    $priorTestDate: String
    $priorTestType: String
    $priorTestResult: String
    $noSymptoms: Boolean
  ) {
    updateTimeOfTestQuestions(
      patientId: $patientId
      pregnancy: $pregnancy
      symptoms: $symptoms
      noSymptoms: $noSymptoms
      firstTest: $firstTest
      priorTestDate: $priorTestDate
      priorTestType: $priorTestType
      priorTestResult: $priorTestResult
      symptomOnset: $symptomOnset
    )
  }
`;

interface AreYouSureProps {
  patientName: string;
  cancelHandler: () => void;
  continueHandler: () => void;
}

const AreYouSure: React.FC<AreYouSureProps> = ({
  patientName,
  cancelHandler,
  continueHandler,
}) => (
  <Modal
    isOpen={true}
    style={{
      content: {
        top: "50%",
        left: "50%",
        width: "40%",
        minWidth: "20em",
        maxHeight: "14em",
        marginRight: "-50%",
        transform: "translate(-50%, -50%)",
      },
    }}
    overlayClassName={"prime-modal-overlay"}
    contentLabel="Questions not answered"
  >
    <p className="usa-prose prime-modal-text">
      Time of test questions for <b> {` ${patientName} `} </b> have not been
      completed. Do you want to submit results anyway?
    </p>
    <div className="prime-modal-buttons">
      <Button onClick={cancelHandler} unstyled label="No, go back" />
      <Button onClick={continueHandler} label="Submit Anyway" />
    </div>
  </Modal>
);
Modal.setAppElement("#root");

interface QueueItemProps {
  internalId: string;
  patient: {
    internalId: string;
    firstName: string;
    middleName: string;
    lastName: string;
    lookupId: string;
    telephone: string;
    birthDate: string;
  };
  devices: {
    name: string;
    internalId: string;
  }[];
  askOnEntry: string;
  selectedDeviceId: string;
  selectedTestResult: TestResult;
  defaultDevice: {
    internalId: string;
  };
  refetchQueue: () => void;
  facilityId: string;
}

interface updateQueueItemProps {
  deviceId?: string;
  result?: TestResult;
}

const QueueItem: any = ({
  internalId,
  patient,
  devices,
  askOnEntry,
  selectedDeviceId,
  selectedTestResult,
  defaultDevice,
  refetchQueue,
  facilityId,
}: QueueItemProps) => {
  const appInsights = useAppInsightsContext();
  const trackRemovePatientFromQueue = useTrackEvent(
    appInsights,
    "Remove Patient From Queue",
    {}
  );
  const trackSubmitTestResult = useTrackEvent(
    appInsights,
    "Submit Test Result",
    {}
  );
  const trackUpdateAoEResponse = useTrackEvent(
    appInsights,
    "Update AoE Response",
    {}
  );

  const [mutationError, updateMutationError] = useState(null);
  const [removePatientFromQueue] = useMutation(REMOVE_PATIENT_FROM_QUEUE);
  const [submitTestResult] = useMutation(SUBMIT_TEST_RESULT);
  const [updateAoe] = useMutation(UPDATE_AOE);
  const [editQueueItem] = useMutation<
    EditQueueItemResponse,
    EditQueueItemParams
  >(EDIT_QUEUE_ITEM);

  const [isAoeModalOpen, updateIsAoeModalOpen] = useState(false);
  const [aoeAnswers, setAoeAnswers] = useState(askOnEntry);

  const [deviceId, updateDeviceId] = useState(
    selectedDeviceId || defaultDevice.internalId
  );
  const [testResultValue, updateTestResultValue] = useState<
    TestResult | undefined
  >(selectedTestResult);

  const [isConfirmationModalOpen, updateIsConfirmationModalOpen] = useState(
    false
  );
  let forceSubmit = false;

  if (mutationError) {
    throw mutationError;
  }

  const testResultsSubmitted = () => {
    let { type, title, body } = {
      ...ALERT_CONTENT[QUEUE_NOTIFICATION_TYPES.SUBMITTED_RESULT__SUCCESS](
        patient
      ),
    };
    let alert = <Alert type={type} title={title} body={body} />;
    showNotification(toast, alert);
  };

  const onTestResultSubmit = (e?: any) => {
    if (e) e.preventDefault();
    if (forceSubmit || areAnswersComplete(aoeAnswers)) {
      trackSubmitTestResult({});
      submitTestResult({
        variables: {
          patientId: patient.internalId,
          deviceId: deviceId,
          result: testResultValue,
        },
      })
        .then(testResultsSubmitted)
        .then(refetchQueue)
        .then(() => removeTimer(internalId))
        .catch((error) => updateMutationError(error));
    } else {
      updateIsConfirmationModalOpen(true);
    }
  };

  const updateQueueItem = ({ deviceId, result }: updateQueueItemProps) => {
    editQueueItem({
      variables: {
        id: internalId,
        deviceId,
        result,
      },
    }).then(
      (response) => {
        if (!response.data) {
          throw Error("null response from update queue");
        }
        updateDeviceId(response.data.editQueueItem.deviceType.internalId);
        updateTestResultValue(response.data.editQueueItem.result);
      },
      (error) => {
        updateMutationError(error);
      }
    );
  };

  const onDeviceChange = (e: React.FormEvent<HTMLSelectElement>) => {
    const deviceId = (e.target as HTMLSelectElement).value;
    updateQueueItem({ deviceId });
  };

  const onTestResultChange = (result: TestResult | undefined) => {
    updateQueueItem({ result });
  };

  const removeFromQueue = (patientId: string) => {
    trackRemovePatientFromQueue({});
    removePatientFromQueue({
      variables: {
        patientId,
      },
    })
      .then(refetchQueue)
      .then(() => removeTimer(internalId))
      .catch((error) => updateMutationError(error));
  };

  const openAoeModal = () => {
    updateIsAoeModalOpen(true);
  };

  const closeAoeModal = () => {
    updateIsAoeModalOpen(false);
  };

  const saveAoeCallback = (answers: any) => {
    setAoeAnswers(answers);
    trackUpdateAoEResponse({});
    updateAoe({
      variables: {
        patientId: patient.internalId,
        noSymptoms: answers.noSymptoms,
        symptoms: answers.symptoms,
        symptomOnset: answers.symptomOnset,
        pregnancy: answers.pregnancy,
        firstTest: answers.firstTest,
        priorTestDate: answers.priorTestDate,
        priorTestType: answers.priorTestType,
        priorTestResult: answers.priorTestResult,
      },
    }).then(
      (_response) => {
        refetchQueue();
      },
      (error) => {
        updateMutationError(error);
      }
    );
  };

  let options = devices.map((device) => ({
    label: device.name,
    value: device.internalId,
  }));

  const patientFullName = displayFullName(
    patient.firstName,
    patient.middleName,
    patient.lastName
  );

  const closeButton = (
    <div
      onClick={() => removeFromQueue(patient.internalId)}
      className="prime-close-button"
    >
      <span className="fa-layers">
        <FontAwesomeIcon icon={"circle"} size="2x" inverse />
        <FontAwesomeIcon icon={"times-circle"} size="2x" />
      </span>
    </div>
  );

  return (
    <React.Fragment>
      <div className="grid-container prime-container prime-queue-item usa-card__container">
        {closeButton}
        <div className="grid-row">
          <div className="tablet:grid-col-9">
            <div className="grid-row prime-test-name usa-card__header">
              <h2>{patientFullName}</h2>
              <TestTimerWidget id={internalId} />
            </div>
            <div className="grid-row usa-card__body">
              <ul className="prime-ul">
                <li className="prime-li">
                  <LabeledText text={patient.lookupId} label="Unique ID" />
                </li>
                <li className="prime-li">
                  <LabeledText text={patient.telephone} label="Phone Number" />
                </li>
                <li className="prime-li">
                  <LabeledText
                    text={moment(patient.birthDate).format("MM/DD/yyyy")}
                    label="Date of Birth"
                  />
                </li>
                <li className="prime-li">
                  <Anchor
                    text="Time of Test Questions"
                    onClick={openAoeModal}
                  />
                  {isAoeModalOpen && (
                    <AoeModalForm
                      saveButtonText="Save"
                      onClose={closeAoeModal}
                      patient={patient}
                      loadState={aoeAnswers}
                      saveCallback={saveAoeCallback}
                      facilityId={facilityId}
                    />
                  )}
                  <p>
                    <AskOnEntryTag aoeAnswers={aoeAnswers} />
                  </p>
                </li>
              </ul>
            </div>
            <div className="grid-row usa-card__footer">
              <form className="usa-form">
                <Dropdown
                  options={options}
                  label="Device"
                  name="testDevice"
                  selectedValue={deviceId}
                  onChange={onDeviceChange}
                />
              </form>
            </div>
          </div>
          <div className="tablet:grid-col-3 prime-test-result">
            {isConfirmationModalOpen && (
              <AreYouSure
                patientName={patientFullName}
                cancelHandler={() => updateIsConfirmationModalOpen(false)}
                continueHandler={() => {
                  forceSubmit = true;
                  onTestResultSubmit();
                }}
              />
            )}
            <TestResultInputForm
              queueItemId={internalId}
              testResultValue={testResultValue}
              onSubmit={onTestResultSubmit}
              onChange={onTestResultChange}
            />
          </div>
        </div>
      </div>
    </React.Fragment>
  );
};

QueueItem.propTypes = {
  patient: patientPropType,
  devices: PropTypes.arrayOf(devicePropType),
};
export default QueueItem;
