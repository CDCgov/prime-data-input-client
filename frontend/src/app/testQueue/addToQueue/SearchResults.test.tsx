import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";

import { Patient } from "../../patients/ManagePatients";
import { TestResult } from "../../testQueue/QueueItem";

import SearchResults from "./SearchResults";

const dummyTest = {
  dateAdded: "2020-01-01",
  result: "NEGATIVE" as TestResult,
  dateTested: "2020-01-01",
  deviceTypeModel: "MegaTester2000",
};

const patients: Patient[] = [
  {
    internalId: "a123",
    firstName: "George",
    middleName: "",
    lastName: "Washington",
    birthDate: "1950-01-01",
    isDeleted: false,
    role: "somerole",
    lastTest: dummyTest,
  },
  {
    internalId: "b456",
    firstName: "Martin",
    middleName: "Luther",
    lastName: "King",
    birthDate: "1950-01-01",
    isDeleted: false,
    role: "somerole",
    lastTest: dummyTest,
  },
  {
    internalId: "c789",
    firstName: "Barack",
    middleName: "Hussein",
    lastName: "Obama",
    birthDate: "1950-01-01",
    isDeleted: false,
    role: "somerole",
    lastTest: dummyTest,
  },
];

describe("SearchResults", () => {
  it("should say 'No Results' for no matches", () => {
    const component = renderer.create(
      <SearchResults
        page="queue"
        patients={[]}
        patientsInQueue={[]}
        onAddToQueue={jest.fn()}
        shouldShowSuggestions={true}
        loading={false}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should show matching results", () => {
    const component = renderer.create(
      <SearchResults
        page="queue"
        patients={patients}
        patientsInQueue={[]}
        onAddToQueue={jest.fn()}
        shouldShowSuggestions={true}
        loading={false}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("links the non-duplicate patient", () => {
    const addToQueue = jest.fn();
    render(
      <SearchResults
        page="queue"
        patients={patients}
        patientsInQueue={["a123", "c789"]}
        onAddToQueue={addToQueue}
        shouldShowSuggestions={true}
        loading={false}
      />
    );

    expect(screen.getAllByText("Test in progress")).toHaveLength(2);
    expect(screen.getAllByText("Begin test")).toHaveLength(1);
  });
});
