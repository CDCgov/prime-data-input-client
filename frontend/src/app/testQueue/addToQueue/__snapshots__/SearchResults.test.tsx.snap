// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults should say 'No Results' for no matches 1`] = `
<div
  className="card-container shadow-3 results-dropdown"
>
  <div
    className="usa-card__body"
  >
    <p>
      No results
    </p>
  </div>
</div>
`;

exports[`SearchResults should show matching results 1`] = `
<div
  className="card-container shadow-3 results-dropdown"
>
  <div
    className="usa-card__body"
  >
    <table
      className="usa-table usa-table--borderless"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Full name
          </th>
          <th
            scope="col"
          >
            Date of birth
          </th>
          <th
            scope="col"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Washington, George
          </td>
          <td />
          <td>
            <button
              className="usa-button usa-button--unstyled"
              onClick={[Function]}
              type="button"
            >
              Begin test
            </button>
          </td>
        </tr>
        <tr>
          <td>
            King, Martin Luther
          </td>
          <td />
          <td>
            <button
              className="usa-button usa-button--unstyled"
              onClick={[Function]}
              type="button"
            >
              Begin test
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Obama, Barack Hussein
          </td>
          <td />
          <td>
            <button
              className="usa-button usa-button--unstyled"
              onClick={[Function]}
              type="button"
            >
              Begin test
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
